govcd/cse_internal.go

```diff
@@ -0,0 +1,240 @@
+	if err != nil {
+		return "", err
+	}
+
+	memoryHealthCheckYaml, err := clusterSettings.generateMachineHealthCheckYaml()
+	if err != nil {
+		return "", err
+	}
+
+	templateArgs := map[string]string{
+		"ClusterName":                 clusterSettings.Name,
+		"TargetNamespace":             clusterSettings.Name + "-ns",
+		"TkrVersion":                  clusterSettings.TkgVersionBundle.TkrVersion,
+		"TkgVersion":                  clusterSettings.TkgVersionBundle.TkgVersion,
+		"UsernameB64":                 base64.StdEncoding.EncodeToString([]byte(clusterSettings.Owner)),
+		"ApiTokenB64":                 base64.StdEncoding.EncodeToString([]byte(clusterSettings.ApiToken)),
+		"PodCidr":                     clusterSettings.PodCidr,
+		"ServiceCidr":                 clusterSettings.ServiceCidr,
+		"VcdSite":                     clusterSettings.VcdUrl,
+		"Org":                         clusterSettings.OrganizationName,
+		"OrgVdc":                      clusterSettings.VdcName,
+		"OrgVdcNetwork":               clusterSettings.NetworkName,
+		"Catalog":                     clusterSettings.CatalogName,
+		"VAppTemplate":                clusterSettings.KubernetesTemplateOvaName,
+		"ControlPlaneSizingPolicy":    clusterSettings.ControlPlane.SizingPolicyName,
+		"ControlPlanePlacementPolicy": clusterSettings.ControlPlane.PlacementPolicyName,
+		"ControlPlaneStorageProfile":  clusterSettings.ControlPlane.StorageProfileName,
+		"ControlPlaneDiskSize":        fmt.Sprintf("%dGi", clusterSettings.ControlPlane.DiskSizeGi),
+		"ControlPlaneMachineCount":    strconv.Itoa(clusterSettings.ControlPlane.MachineCount),
+		"ControlPlaneEndpoint":        clusterSettings.ControlPlane.Ip,
+		"DnsVersion":                  clusterSettings.TkgVersionBundle.CoreDnsVersion,
+		"EtcdVersion":                 clusterSettings.TkgVersionBundle.EtcdVersion,
+		"ContainerRegistryUrl":        clusterSettings.VcdKeConfig.ContainerRegistryUrl,
+		"KubernetesVersion":           clusterSettings.TkgVersionBundle.KubernetesVersion,
+		"SshPublicKey":                clusterSettings.SshPublicKey,
+		"VirtualIpSubnet":             clusterSettings.VirtualIpSubnet,
+	}
+
+	buf := &bytes.Buffer{}
+	if err := capiYaml.Execute(buf, templateArgs); err != nil {
+		return "", fmt.Errorf("could not generate a correct CAPI YAML: %s", err)
+	}
+
+	// The final "pretty" YAML. To embed it in the final payload it must be marshaled into a one-line JSON string
+	prettyYaml := ""
+	if memoryHealthCheckYaml != "" {
+		prettyYaml += fmt.Sprintf("%s\n---\n", memoryHealthCheckYaml)
+	}
+	prettyYaml += fmt.Sprintf("%s\n---\n%s", nodePoolYaml, buf.String())
+
+	// We don't use a standard json.Marshal() as the YAML contains special characters that are not encoded properly, such as '<'.
+	buf.Reset()
+	enc := json.NewEncoder(buf)
+	enc.SetEscapeHTML(false)
+	err = enc.Encode(prettyYaml)
+	if err != nil {
+		return "", fmt.Errorf("could not encode the CAPI YAML into a JSON string: %s", err)
```

Comment:
Does the `sigs.k8s.io/yaml` not help here?

Reply:
Dainius is probably thinking of `yaml.YAMLToJSON`

Reply:
Yeah, that would be the way to go if we aimed to transform the YAML into JSON and set the (hypothetical) `capiJson` object with it, something like:
```
{
  "apiVersion": "capvcd.vmware.com/v1.1",
  "kind": "CAPVCDCluster",
  "name": "my-cluster",
  "capiJson": {
     -- The nicely JSON object here
  }
}
```
But here is a bit different, as we are "embedding" a one-liner-YAML as a string value for a JSON field:

```
{
  "apiVersion": "capvcd.vmware.com/v1.1",
  "kind": "CAPVCDCluster",
  "name": "my-cluster",
  "capiYaml": "The big YAML chunk with special characters"
}
```

So in our case we must omit that it's a YAML and consider it as a regular string with odd characters that needs to be inserted in a JSON.